pkg/frontend/v2/frontend_scheduler_adapter.go

```diff
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package v2
+
+import (
+	"context"
+	"time"
+
+	"github.com/grafana/dskit/httpgrpc"
+	"github.com/grafana/dskit/tenant"
+
+	"github.com/grafana/mimir/pkg/frontend/querymiddleware"
+	v1 "github.com/grafana/mimir/pkg/frontend/v1"
+	"github.com/grafana/mimir/pkg/querier"
+	"github.com/grafana/mimir/pkg/scheduler/schedulerpb"
+	"github.com/grafana/mimir/pkg/util/validation"
+)
+
+type frontendToSchedulerAdapter struct {
+	cfg             Config
+	limits          v1.Limits
+	prometheusCodec querymiddleware.Codec
+}
+
+func (a *frontendToSchedulerAdapter) frontendToSchedulerEnqueueRequest(
+	req *frontendRequest, frontendAddr string,
+) *schedulerpb.FrontendToScheduler {
+	return &schedulerpb.FrontendToScheduler{
+		Type:                      schedulerpb.ENQUEUE,
+		QueryID:                   req.queryID,
+		UserID:                    req.userID,
+		HttpRequest:               req.request,
+		FrontendAddress:           frontendAddr,
+		StatsEnabled:              req.statsEnabled,
+		AdditionalQueueDimensions: a.extractAdditionalQueueDimensions(req.ctx, req.request),
+	}
+}
+
+const ShouldQueryIngestersQueueDimension = "ingester"
+const ShouldQueryStoreGatewayQueueDimension = "store-gateway"
+const ShouldQueryIngestersAndStoreGatewayQueueDimension = "ingester|store-gateway"
+
+func (a *frontendToSchedulerAdapter) extractAdditionalQueueDimensions(
+	ctx context.Context, request *httpgrpc.HTTPRequest,
+) []string {
+	httpRequest, err := httpgrpc.ToHTTPRequest(ctx, request)
+	if err != nil {
+		return nil
+	}
+	promRequest, err := a.prometheusCodec.DecodeRequest(ctx, httpRequest)
```

Comment:
Both `httpgrpc.ToHTTPRequest` and `prometheusCodec.DecodeRequest` would have been already called successfully higher in the stack in the various middlewares and roundtrippers, so we would not expect failures here.

Whether to return err or return nil in an error case here is matter of whether we want to be stricter or looser.

A failure of this method means the request does not receive an additional queue dimensions, _which the scheduler can handle without issue_. But we could still opt to be stricter here